File: exercises/python/checkout/receipt.py
class Receipt:
    def __init__(self):
        self.text = ""
        self.total = 0
        self.number_of_a = 0
        self.number_of_b = 0

    def scanned_a(self):
        self.text += "A: 50"
        self.number_of_a += 1
        self.total += 50
        if self.number_of_a % 3 == 0:
            self.total -= 20
            self.text += " - 20 (3 for 130)"
        self.text += "\n"

    def scanned_b(self):
        self.text += "B: 30"
        self.number_of_b += 1
        self.total += 30
        if self.number_of_b % 2 == 0:
            self.total -= 15
            self.text += " - 15 (2 for 45)"
        self.text += "\n"

    def scanned_c(self):
        self.text += "C: 20\n"
        self.total += 20

    def scanned_d(self):
        self.text += "D: 15\n"
        self.total += 15

    def get_text(self):
        return '{text}Total: {total}'.format(text=self.text, total=self.total)

File: exercises/java/src/main/java/checkout/Receipt.java
package checkout;

class Receipt {

    private String text = "";
    private int numberOfA = 0;
    private int numberOfB = 0;
    private int numberOfC = 0;
    private int numberOfD = 0;
    private int total;

    String text() {
        return text + "Total: " + total;
    }

    void scannedA() {
        scannedItem("A", 50, ++numberOfA, 5, 220);
    }

    void scannedB() {
        scannedItem("B", 30, ++numberOfB, 2, 45);
    }

    void scannedC() {
        scannedItem("C", 20, ++numberOfC, 4, 70);
    }

    void scannedD() {
        scannedItem("D", 15, ++numberOfD, 5, 60);
    }

    private void scannedItem(String itemName, int price, int numberScanned, int numberOfItems, int discountedAmount) {
        int discount =  price * numberOfItems - discountedAmount;
        text += itemName + ": " + price;
        if (numberScanned % numberOfItems == 0) {
            text += " - " + discount + " (" + numberOfItems + " for " + discountedAmount + ")";
            this.total -= discount;
        }
        total += price;
        text += "\n";
    }
}
